store.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2018 Canonical Ltd.
+// Licensed under the LGPL, see LICENCE file for details.
+
+package aclstore
+
+import (
+	"context"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/juju/simplekv"
+	"gopkg.in/errgo.v1"
+)
+
+var ErrACLNotFound = errgo.Newf("ACL not found")
+
+// separator is used as the character to divide usernames in the ACL.
+// This needs to be a character that's illegal in usernames.
+const separator = "\n"
+
+// ACLStore is the persistent storage interface used by an ACLHandler.
+type ACLStore interface {
+	// NewACL creates an ACL with the given name and initial users.
+	// If the ACL already exists, this is a no-op and the initialUsers
+	// argument is ignored.
+	NewACL(ctx context.Context, aclName string, initialUsers ...string) error
+
+	// Add adds users to the ACL with the given name.
+	// Adding a user that's already in the ACL is a no-op.
+	// It returns an error with an ErrACLNotFound cause if the ACL
+	// does not exist.
+	Add(ctx context.Context, aclName string, users ...string) error
+
+	// Remove removes users from the ACL with the given name.
+	// It returns an error with an ErrACLNotFound cause if the ACL
+	// does not exist. It returns an error with an ErrUserNotFound
+	// cause if any of the users do not exist.
+	// TODO should it do nothing in that case?
+	Remove(ctx context.Context, aclName string, users ...string) error
+
+	// Set sets the users held in the ACL with the given name.
+	// It returns an ErrACLNotFound cause if the ACL does not
+	// exist.
+	Set(ctx context.Context, aclName string, users ...string) error
+
+	// Get returns the users held in the ACL with the given name,
+	// sorted lexically. It returns an error with an ErrACLNotFound cause
+	// if the ACL does not exist.
+	Get(ctx context.Context, aclName string) ([]string, error)
+}
+
+// NewACLStore returns an ACLStore implementation that uses an underlying
+// key-value store for persistent storage.
+func NewACLStore(kv simplekv.Store) ACLStore {
+	return &kvStore{kv}
+}
+
+type kvStore struct {
+	kv simplekv.Store
+}
+
+var errAlreadyExists = errgo.Newf("ACL already exists")
+
+// NewACL implements ACLStore.NewACL.
+func (s *kvStore) NewACL(ctx context.Context, aclName string, initialUsers ...string) error {
+	err := s.kv.Update(ctx, aclName, time.Time{}, func(val []byte) ([]byte, error) {
+		if val != nil {
+			return nil, errAlreadyExists
+		}
+		newVal, err := s.aclToValue(initialUsers)
+		if err != nil {
+			return nil, errgo.Mask(err)
+		}
+		return newVal, nil
+	})
+	if err != nil {
```

Comment:
as errgo.Mask(nil) returns nil you can simplify with just the if content ?

Reply:
With the extra logic here, I'm not sure it's clearer to do that. Further down, perhaps it might be. I'll leave it for now.